### PR TITLE
Also prevent too much recursion in the assembly parser.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: test/libsolidity/InlineAssembly.cpp [hunk 1]
```diff
@@ -392,6 +392,20 @@ BOOST_AUTO_TEST_CASE(instruction_too_many_arguments)
 	CHECK_PARSE_ERROR("{ mul(1, 2, 3) }", ParserError, "Expected ')' (\"mul\" expects 2 arguments)");
 }
 
+BOOST_AUTO_TEST_CASE(recursion_depth)
+{
+	string input;
+	for (size_t i = 0; i < 20000; i++)
+		input += "{";
+	input += "let x:u256 := 0:u256";
```
Comment (axic) [inline test/libsolidity/InlineAssembly.cpp]: Should not have types here.

**Comment by axic** [comment]
Include in changelog please.

**Comment by chriseth** [comment]
Fixed.

File: test/libsolidity/InlineAssembly.cpp (hunk 1 above)
Comment (axic) [reply test/libsolidity/InlineAssembly.cpp]: Why are the tests passing if this has types?

Comment (chriseth) [reply test/libsolidity/InlineAssembly.cpp]: Because it requires there to be an error. I guess the parser just continues after the `:`.

Comment (chriseth) [reply test/libsolidity/InlineAssembly.cpp]: Ah right, it does not get until that point.

**Comment by axic** [comment]
I get this:

```
unknown location:0: fatal error: in "SolidityInlineAssembly/Parsing/recursion_depth": memory access violation at address: 0x7fff5079ad60: invalid permissions
/Users/alex/Projects/solidity/test/libsolidity/InlineAssembly.cpp:403: last checkpoint: "recursion_depth" entry.
```

```
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x7fff5f3ffc80)
    frame #0: 0x0000000101860c99 libsolidity.dylib`dev::solidity::Scanner::scanToken() + 25
libsolidity.dylib`dev::solidity::Scanner::scanToken:
->  0x101860c99 <+25>: movq   %rdi, -0x8e0(%rbp)
    0x101860ca0 <+32>: movq   -0x8e0(%rbp), %rax
    0x101860ca7 <+39>: movq   %rax, %rdi
    0x101860caa <+42>: addq   $0xc0, %rdi
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=2, address=0x7fff5f3ffc80)
  * frame #0: 0x0000000101860c99 libsolidity.dylib`dev::solidity::Scanner::scanToken() + 25
    frame #1: 0x0000000101862587 libsolidity.dylib`dev::solidity::Scanner::next() + 87
    frame #2: 0x000000010185fffa libsolidity.dylib`dev::solidity::ParserBase::expectToken(dev::solidity::Token::Value) + 14378
    frame #3: 0x00000001018a00de libsolidity.dylib`dev::solidity::assembly::Parser::parseBlock() + 158
    frame #4: 0x00000001018a098b libsolidity.dylib`dev::solidity::assembly::Parser::parseStatement() + 523
    frame #5: 0x00000001018a0133 libsolidity.dylib`dev::solidity::assembly::Parser::parseBlock() + 243
...
    frame #3454: 0x00000001018a098b libsolidity.dylib`dev::solidity::assembly::Parser::parseStatement() + 523
    frame #3455: 0x00000001018a0133 libsolidity.dylib`dev::solidity::assembly::Parser::parseBlock() + 243
    frame #3456: 0x00000001018a098b libsolidity.dylib`dev::solidity::assembly::Parser::parseStatement() + 523
    frame #3457: 0x00000001018a0133 libsolidity.dylib`dev::solidity::assembly::Parser::parseBlock() + 243
    frame #3458: 0x000000010189fec3 libsolidity.dylib`dev::solidity::assembly::Parser::parse(std::__1::shared_ptr<dev::solidity::Scanner> const&) + 483
    frame #3459: 0x00000001017413ba libsolidity.dylib`dev::solidity::AssemblyStack::parseAndAnalyze(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 1034
    frame #3460: 0x000000010027e93f soltest`dev::solidity::test::(anonymous namespace)::parseAndReturnFirstError(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool, dev::solidity::AssemblyStack::Machine) + 319
    frame #3461: 0x00000001002444c8 soltest`dev::solidity::test::(anonymous namespace)::expectError(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, bool, bool) + 136
    frame #3462: 0x000000010026551a soltest`dev::solidity::test::SolidityInlineAssembly::Parsing::recursion_depth::test_method() + 586
    frame #3463: 0x00000001002650f5 soltest`dev::solidity::test::SolidityInlineAssembly::Parsing::recursion_depth_invoker() + 677
...
```

File: test/libsolidity/InlineAssembly.cpp (hunk 1 above)
Comment (axic) [reply test/libsolidity/InlineAssembly.cpp]: Yep, 20000 is over the limit.

**Comment by chriseth** [comment]
Interesting! Yes, macos has a much smaller stack size. So I guess we have to reduce the limit or reduce the size of stack frames.

**Comment by chriseth** [comment]
Reduced the max depth to 3000.

**Comment by axic** [comment]
There must be a different problem to stack size. The test for the Solidity parser works (worked even with the old limit), but the inline assembly one still fails even after reducing the stack limit.

**Comment by chriseth** [comment]
I think this is because the inline assembly parser just has larger stack frames. The inline assembly AST is built on the stack while the Solidity AST is built on the heap.

**Comment by chriseth** [comment]
Of course the proper way is to not have any recursion at all.

**Comment by axic** [comment]
Ran it again and it works. Probably there was some caching issue before.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
I guess before merging we should run the test in solc-js.

**Comment by chriseth** [comment]
Ok, will run the tests.

**Comment by chriseth** [comment]
Tests pass.

**Comment by axic** [comment]
> I guess before merging we should run the test in solc-js.

With this I meant add a test case to solc-js and run it, since we don't know what kind of stack size / stack depth Emscripten / JS allows.

**Comment by chriseth** [comment]
Are you talking about a test that checks whether the recursion limit is hit before the stack height limit is hit?

**Comment by axic** [comment]
Yes, the same test we have here, but for solc-js (to validate Emscripten).

**Pull request merged**